src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -474,58 +482,86 @@ private static bool IsValidIPv6Host(string hostText)
                     // [::1] is the shortest valid IPv6 host
                     if (i < 4)
                     {
-                        return false;
+                        BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
+                    }
+                    else
+                    {
+                        // Tail call
+                        ValidateHostPort(hostText, i + 1);
+                        return;
                     }
-                    return IsValidHostPort(hostText, i + 1);
                 }
 
                 if (!IsHex(ch) && ch != ':' && ch != '.')
                 {
-                    return false;
+                    BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
                 }
             }
 
             // Must contain a ']'
-            return false;
+            BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static bool IsValidHostPort(string hostText, int offset)
+        private static void ValidateHostPort(string hostText, int offset)
         {
-            if (offset == hostText.Length)
+            // Skip bounds check for accessing the [offset] element
+            if ((uint)offset >= (uint)hostText.Length)
```

Comment:
So the bounds check get repeated *unless* you cat offset to a uint? If so, that seems crazy. It seemed odd when the offset started as a char and it even seems more odd now that the offset is the int.

Reply:
You have to check its not less than zero and also not greater than or equal to length; so two tests. Which is one test with the `uint`. Also I think if you do it in two tests, it doesn't work - outside of for loops with a constant start.